### PR TITLE
Hosting Type Providers out-of-process

### PR DESCRIPTION
**Pull request by DedSec256**
```c# 
"Hello from the in-process type provider!".Replace("in-process", "out-of-process")
```

ezgif com-gif-maker

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
image

File: rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+    private val ILModuleRef = structdef {
+        field("Name", string)
+        field("HasMetadata", bool)
+        field("Hash", array(byte))
+    }
+
+    private val ILScopeRef = structdef {
+        field("IsModuleRef", bool)
+        field("IsAssemblyRef", bool)
```
Comment (auduchinok) [inline rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt]: Maybe use enum instead?

File: rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        field("parameters", string)
+    }
+
+    private val GetGeneratedAssemblyContentsParameters = structdef {
+        field("typeProviderId", int)
+        field("syntheticMethodBase", string)
+        field("parameters", string)
+    }
+
+    private val pos = structdef {
```
Comment (auduchinok) [inline rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt]: Let's use `Rd` prefix for protocol types. It'll be something like `RdPos` here.

File: rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+import com.jetbrains.rd.generator.nova.*
+import com.jetbrains.rd.generator.nova.PredefinedType.*
+import com.jetbrains.rd.generator.nova.csharp.CSharp50Generator
+
+@Suppress("unused")
+object RdFSharpTypeProvidersLoaderModel : Root(CSharp50Generator(
+        FlowTransform.Symmetric,
+        "JetBrains.ReSharper.Features.WinForms.Designer.Protocol.Model")
+) {
+    private val Type = structdef {
```
Comment (auduchinok) [inline rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt]: `RdProvidedType`?
You then won't have to do stuff like `using Type = System.Type;`.

File: rider-fsharp/src/main/java/com/jetbrains/rider/plugins/fsharp/services/fsi/FsiConsoleRunner.kt [hunk 4]
```diff
@@ -157,7 +158,7 @@ class FsiConsoleRunner(sessionInfo: RdFsiSessionInfo, val fsiHost: FsiHost, debu
             commandHistory.addEntry(CommandHistory.Entry(visibleText, fsiText))
 
             // show the window without getting focus
-            ExecutionManager.getInstance(project).contentManager.selectRunContent(contentDescriptor)
+            RunContentManager.getInstance(project).selectRunContent(contentDescriptor)
```
Comment (auduchinok) [inline rider-fsharp/src/main/java/com/jetbrains/rider/plugins/fsharp/services/fsi/FsiConsoleRunner.kt]: Please rebase.

**Comment by DedSec256** [comment]
@auduchinok 
image

File: ReSharper.FSharp/src/TypeProvidersLoader/TypeProvidersLoader.exe.config [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<configuration>
+    <runtime>
+      <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
+        <dependentAssembly>
+          <assemblyIdentity name="FSharp.Core" publicKeyToken="b03f5f7f11d50a3a" culture="neutral"/>
+          <bindingRedirect oldVersion="2.0.0.0-9.9.9.9" newVersion="4.6.0.0"/>
```
Comment (auduchinok) [inline ReSharper.FSharp/src/TypeProvidersLoader/TypeProvidersLoader.exe.config]: JFYI: we've migrated to `4.7.0.0` in `net201`, this file will need to be updated after rebase.

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+                            None
+
+                    let newTypeProviders = 
+                        [ match designTimeAssemblyName, resolutionEnvironment.outputFile with
+                            | Some designTimeAssemblyName, Some path when String.Compare(designTimeAssemblyName.Name,
+                                                                                         Path.GetFileNameWithoutExtension path,
+                                                                                         StringComparison.OrdinalIgnoreCase) = 0 -> ()
+                            | Some _, _ ->
+                            let res = ourModel.InstantiateTypeProvidersOfAssembly.Sync(InstantiateTypeProvidersOfAssemblyParameters(
+                                                                                                                                  runTimeAssemblyFileName,
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs]: Please reformat.

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+            providedTypes
+            
+        member this.ResolveTypeName(pn: Tainted<IProvidedNamespace>, typeName: string, m: range) =
+            pn.PApply((fun providedNamespace ->
+                (providedNamespace.As<IProxyProvidedNamespace>().ResolveProvidedTypeName typeName).WithCache()), range=m) 
+
+            
+    interface IDisposable with
+        member this.Dispose() = () //terminate connection
+
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs]: Does this file have correct line separators? This last line looks suspicious.

File: rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+
+    private val RdProvidedMethodInfoProcessModel = aggregatedef("RdProvidedMethodInfoProcessModel") {
+        call("DeclaringType", int, int.nullable)
+        call("ReturnType", int, int)
+        call("GetParameters", int, array(RdProvidedParameterInfo))
+        call("GetGenericArguments", int, array(int))
+    }
+
+    private val InstantiationResult = structdef {
+        field("IsSuccsses", bool)
```
Comment (auduchinok) [inline rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt]: Typo.

File: rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+        call("GetTypes", int, array(int))
+        call("ResolveTypeName", structdef("ResolveTypeNameArgs") {
+            field("Id", int)
+            field("TypeFullName", string)
+        }, int)
+    }
+
+    private val RdProvidedNamespace = classdef {
+        field("NamespaceName", string)
+        field("EntityId", int)
```
Comment (auduchinok) [inline rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt]: Could it be just `Id` like in some other models? And, well, let's use a single word across these types. :)

File: rider-fsharp/build.gradle.kts [hunk 10]
```diff
@@ -112,6 +112,7 @@ val pluginFiles = listOf(
         "FSharp.ProjectModelBase/bin/$buildConfiguration/net461/JetBrains.ReSharper.Plugins.FSharp.ProjectModelBase",
         "FSharp.Common/bin/$buildConfiguration/net461/JetBrains.ReSharper.Plugins.FSharp.Common",
         "FSharp.Psi/bin/$buildConfiguration/net461/JetBrains.ReSharper.Plugins.FSharp.Psi",
+		"FSharp.TypeProvidersProtocol/bin/$buildConfiguration/net461/JetBrains.ReSharper.Plugins.FSharp.TypeProvidersProtocol",
```
Comment (auduchinok) [inline rider-fsharp/build.gradle.kts]: Please fix the indentation.

File: rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt (hunk 3 above)
Comment (DedSec256) [reply rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt]: Fixed

File: ReSharper.FSharp/src/TypeProvidersLoader/TypeProvidersLoader.exe.config (hunk 5 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/TypeProvidersLoader/TypeProvidersLoader.exe.config]: Fixed

File: rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt (hunk 1 above)
Comment (DedSec256) [reply rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt]: > Please use enum instead.

corrected for models in which the total size of boolean fields is more than 4 bytes (sizeof(int/enum))

File: rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt (hunk 2 above)
Comment (DedSec256) [reply rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt]: > Let's use `Rd` prefix for protocol types. It'll be something like `RdPos` here.

fixed

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+           (this :> IDisposable).Dispose()
+           ExtensionTypingProvider <- defaultExtensionTypingProvider) |> ignore
+       ()
+       
+    let onInitialized lifetime model =
+        ourModel <- model
+        ()
+        
+    let onFailed() = ()
+    let typeProvidersCache: IDictionary<string, ITypeProvider list> = Dictionary<string, ITypeProvider list>() :> _
```
[the remark this answers is not in the document]
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs:37]: > `Dictionary` type arguments can be omitted here.

fixed

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+    
+    interface IExtensionTypingProvider with
+        member this.InstantiateTypeProvidersOfAssembly
+                    (runTimeAssemblyFileName: string, 
+                     ilScopeRefOfRuntimeAssembly: ILScopeRef, 
+                     designTimeAssemblyNameString: string, 
+                     resolutionEnvironment: ResolutionEnvironment, 
+                     isInvalidationSupported: bool, 
+                     isInteractive: bool, 
+                     systemRuntimeContainsType: string -> bool, //не забыть
```
[the remark this answers is not in the document]
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs]: > Please use `todo:` comments instead: it'll be significanlty easier to find all places needing further work later.

fixed

**Comment by auduchinok** [comment]
Please rebase to `net202`.

File: ReSharper.FSharp/test/src/FSharp.Tests/TypeProviders/ProvidedPropertiesResolveTest.fs [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+    
+    override x.RelativeTestDataPath = "features/typeProviders/providedProperties"
+
+    [<Test>] member x.``Readonly string property``() = x.DoNamedTest()
+    [<Test>] member x.``Readonly internal type property``() = x.DoNamedTest()
+    [<Test>] member x.``Static readonly string property``() = x.DoNamedTest()
+    [<Test>] member x.``Try set readonly property``() = x.DoNamedTest()
+    [<Test>] member x.``String property with setter``() = x.DoNamedTest()
+    [<Test>] member x.``String property with indexer``() = x.DoNamedTest()
+    [<Test>] member x.``Set mutable property``() = x.DoNamedTest()
```
Comment (DedSec256) [inline ReSharper.FSharp/test/src/FSharp.Tests/TypeProviders/ProvidedPropertiesResolveTest.fs:18]: bruh...

File: ReSharper.FSharp/test/src/FSharp.Tests/TypeProviders/ProvidedEventResolveTest.fs [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+open NUnit.Framework
+
+[<TestReferences("../../../assemblies/TestTypeProviders/net45/TestTypeProviders.dll")>]
+type ProvidedEventsResolveTest() =
+    inherit TypeProvidersHighlightingTestBase()
+    
+    override x.RelativeTestDataPath = "features/typeProviders/providedEvents"
+
+    [<Test>] member x.``Simple event``() = x.DoNamedTest()
+    [<Test>] member x.``Simple static event``() = x.DoNamedTest()
```
Comment (DedSec256) [inline ReSharper.FSharp/test/src/FSharp.Tests/TypeProviders/ProvidedEventResolveTest.fs:13]: Bruh...

File: ReSharper.FSharp/test/src/FSharp.Tests/TypeProviders/RegexProviderTests.fs [hunk 15]
```diff
@@ -0,0 +1,13 @@
+﻿namespace JetBrains.ReSharper.Plugins.FSharp.Tests.Features.TypeProviders
+
+open JetBrains.ReSharper.TestFramework
+open NUnit.Framework
+
+[<TestReferences(Paths.RegexProviderDll); TestPackages("System.Text.RegularExpressions")>]
```
Comment (auduchinok) [inline ReSharper.FSharp/test/src/FSharp.Tests/TypeProviders/RegexProviderTests.fs:6]: There should be no need to commit its binary, I believe it should be possible to reference the provider package from NuGet. Does `[<TestPackages>]` work for you?

**Comment by varon** [comment]
An enormous thank you for the work done so far on this. Myself and many others are looking forward to the day this is merged!

**Review by xtmq — APPROVED** [review]

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/ZoneMarker.fs [hunk 16]
```diff
@@ -5,4 +5,5 @@ open JetBrains.ReSharper.Plugins.FSharp
 
 [<ZoneMarker>]
 type ZoneMarker() =
-    interface IRequire<ILanguageFSharpZone>
\ No newline at end of file
+    let a = [1, 2]
```
Comment (ForNeVeR) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/ZoneMarker.fs:8]: Looks like a temporary leftover.

File: ReSharper.FSharp/src/FSharp.TypeProvidersProtocol/src/Cache/ProvidedAssembliesCache.cs [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    protected override ProvidedAssembly Create(int key, int typeProviderId, ProvidedTypeContextHolder context)
+      => ProxyProvidedAssembly.Create(
+        TypeProvidersContext.Connection.ExecuteWithCatch(() =>
+          ProvidedAssembliesProcessModel.GetProvidedAssembly.Sync(key)),
+        TypeProvidersContext.Connection);
+
+    protected override ProvidedAssembly[] CreateBatch(int[] keys, int typeProviderId, ProvidedTypeContextHolder context)
+      => throw new System.NotImplementedException();
```
Comment (ForNeVeR) [inline ReSharper.FSharp/src/FSharp.TypeProvidersProtocol/src/Cache/ProvidedAssembliesCache.cs:25]: _(Minor)_ `NotSupportedException` is better for this case, I believe. `NotImplementedException` is more like a TODO (which is appropriate if we plan to implement this method one day, which I don't think is the case here).

File: ReSharper.FSharp/src/FSharp.TypeProvidersProtocol/src/Utils/ProvidedTypesComparer.cs [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+using static FSharp.Compiler.ExtensionTyping;
+
+namespace JetBrains.ReSharper.Plugins.FSharp.TypeProvidersProtocol.Utils
+{
+  public class ProvidedTypesComparer : IEqualityComparer<ProvidedType>
+  {
+    private ProvidedTypesComparer()
+    {
+    }
+    //TODO: check and comment
```
Comment (ForNeVeR) [inline ReSharper.FSharp/src/FSharp.TypeProvidersProtocol/src/Utils/ProvidedTypesComparer.cs:11]: Yep. I am almost convinced this implementation is good enough for our purposes, but please add a comment.

File: ReSharper.FSharp/src/TypeProvidersLoader/Protocol/ModelCreators/ProvidedRdModelsCreatorWithCacheBase.cs [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+    {
+      if (providedModel == null) return default;
+
+      var id = CreateEntityKey(providedModel);
+
+      myCache.Add(id, (providedModel, typeProviderId));
+      return id;
+    }
+
+    //TODO: assert threading
```
Comment (ForNeVeR) [inline ReSharper.FSharp/src/TypeProvidersLoader/Protocol/ModelCreators/ProvidedRdModelsCreatorWithCacheBase.cs:38]: For the record: we've agreed that we don't want to assert it everywhere, but it would be cool to at least guard this increment. If easily achievable, of course.

File: ReSharper.FSharp/src/TypeProvidersLoader/Protocol/ModelCreators/ProvidedTypeCreator.cs [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+      var isValueType = logger.Catch(() => !providedModel.IsGenericParameter && providedModel.IsValueType);
+      var isClass = logger.Catch(() => !providedModel.IsGenericParameter && providedModel.IsClass);
+
+      var baseTypeId = !providedModel.IsGenericParameter
+        ? GetOrCreateId(providedModel.BaseType, typeProviderId)
+        : default;
+
+      var declaringTypeId = providedModel.IsGenericParameter ||
+                            providedModel.IsArray && providedModel.GetElementType().IsGenericParameter
+        ? default
```
Comment (ForNeVeR) [inline ReSharper.FSharp/src/TypeProvidersLoader/Protocol/ModelCreators/ProvidedTypeCreator.cs:29]: Is this `default` just a fancy way of writing `0`? :)

File: ReSharper.FSharp/test/data/assemblies/LemonadeProvider/src/TestTypeProviders/TestTypeProviders.fsproj [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+    </Compile>
+    <Compile Include="AssemblyInfo.fs" />
+    <Compile Include="TestTypeProviders.fs" />
+    <None Include="paket.references" />
+    <None Include="..\..\packages\NETStandard.Library.NETFramework\build\net461\lib\netstandard.dll" Condition="'$(TargetFramework)' == 'net45'">
+        <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+  </ItemGroup>
+  <Import Project="..\..\.paket\Paket.Restore.targets" />
+</Project>
```
Comment (ForNeVeR) [inline ReSharper.FSharp/test/data/assemblies/LemonadeProvider/src/TestTypeProviders/TestTypeProviders.fsproj:34]: I have a question regarding the test data: are you sure we have to commit `*.dll` file shere? /cc @auduchinok

File: rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt [hunk 22]
```diff
@@ -0,0 +1,373 @@
...
+        }
+
+        field("RdTypeProviderProcessModel", RdTypeProviderProcessModel)
+        field("RdProvidedTypeProcessModel", RdProvidedTypeProcessModel)
+        field("RdProvidedMethodInfoProcessModel", RdProvidedMethodInfoProcessModel)
+        field("RdProvidedConstructorInfoProcessModel", RdProvidedConstructorInfoProcessModel)
+        field("RdProvidedAssemblyProcessModel", RdProvidedAssemblyProcessModel)
+        field("RdTestHost", RdTestHost)
+    }
+}
```
Comment (ForNeVeR) [inline rider-fsharp/protocol/src/kotlin/model/RdFSharpTypeProvidersLoaderModel.kt:373]: _(Minor)_ Some of the field names in this file start from uppercase letter, and some from a lowercase one. It would be nice to have single standard (since the generator will fine-tune it for target languages anyway).

File: rider-fsharp/testData/solutions/TypeProviderLibrary/gold/simpleErasedProvider.tmp [hunk 23]
```diff
@@ -0,0 +1,145 @@
+Type Providers:
```
Comment (ForNeVeR) [inline rider-fsharp/testData/solutions/TypeProviderLibrary/gold/simpleErasedProvider.tmp:1]: Should we commit these `*.tmp` files?

**Review by ForNeVeR — CHANGES_REQUESTED** [review]
Overall: a very solid work. From my side, good to go after a few small changes.

File: ReSharper.FSharp/test/data/assemblies/LemonadeProvider/src/TestTypeProviders/TestTypeProviders.fsproj (hunk 21 above)
Comment (auduchinok) [reply ReSharper.FSharp/test/data/assemblies/LemonadeProvider/src/TestTypeProviders/TestTypeProviders.fsproj:34]: We do currently commit few small test assemblies. We're going to need more in future, so the plan is to compile them before/during running the tests.

Comment (ForNeVeR) [reply ReSharper.FSharp/test/data/assemblies/LemonadeProvider/src/TestTypeProviders/TestTypeProviders.fsproj:34]: So, for now this is in line with established practice. Thanks for clarification. Okay then!

File: ReSharper.FSharp/src/FSharp.TypeProvidersProtocol/src/Models/ProxyTypeProvider.cs [hunk 24]
```diff
@@ -70,7 +70,6 @@ public void OnInvalidate()
     public ProvidedExpr GetInvokerExpression(ProvidedMethodBase methodBase, ProvidedVar[] paramExprs) =>
       methodBase switch
       {
-        //TODO: single
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.TypeProvidersProtocol/src/Models/ProxyTypeProvider.cs]: What's the resolution of this todo?

File: ReSharper.FSharp/Directory.Build.props [hunk 25]
```diff
@@ -12,6 +12,7 @@
     <FSharpCoreVersion>5.0.0</FSharpCoreVersion>
     <FantomasVersion>2020.3.0</FantomasVersion>
     <ValueTupleVersion>4.5.0</ValueTupleVersion>
+    <CSharpLanguageVersion>8</CSharpLanguageVersion>
```
Comment (auduchinok) [inline ReSharper.FSharp/Directory.Build.props:15]: Please don't put it to the group that specifies package versions, it becomes a mess.

File: ReSharper.FSharp/ReSharper.FSharp.sln.DotSettings.user [hunk 26]
```diff
@@ -0,0 +1,2 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
+	<s:String x:Key="/Default/Environment/Hierarchy/Build/BuildTool/CustomBuildToolPath/@EntryValue">C:\Program Files\dotnet\sdk\5.0.100\MSBuild.dll</s:String></wpf:ResourceDictionary>
```
Comment (auduchinok) [inline ReSharper.FSharp/ReSharper.FSharp.sln.DotSettings.user:2]: Why do we need this?

File: ReSharper.FSharp/src/FSharp.Common/FSharp.Common.fsproj [hunk 27]
```diff
@@ -13,6 +13,7 @@
     <Compile Include="src\Util\FSharpRangeUtil.fs" />
     <Compile Include="src\Util\FSharpCompilerAttributesUtil.fs" />
     <Compile Include="src\Util\FSharpAssemblyUtil.fs" />
+    <Compile Include="src\Util\FSharpConverter.fs" />
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/FSharp.Common.fsproj:16]: Can it be a more meaningful name, please?

File: ReSharper.FSharp/src/FSharp.Common/FSharp.Common.fsproj [hunk 28]
```diff
@@ -47,6 +48,9 @@
     <Compile Include="src\Checker\ScriptFcsProjectProvider.fs" />
     <Compile Include="src\Checker\FcsProjectBuilder.fs" />
     <Compile Include="src\Checker\FcsProjectProvider.fs" />
+    <Compile Include="src\Shim\TypeProviders\TcImportsWeakHack.fs" />
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/FSharp.Common.fsproj:51]: `WeakHack` looks very wrong to me.

File: ReSharper.FSharp/src/FSharp.Common/FSharp.Common.fsproj [hunk 29]
```diff
@@ -60,12 +64,12 @@
   </PropertyGroup>
 
   <Target Name="CopyFSharpCoreXmlDoc" AfterTargets="AfterBuild">
-    <Copy SourceFiles="$([System.IO.Path]::Combine($(NugetPackageRoot), $(FSharpCoreXmlDocRelativePath)))"
-          DestinationFolder="$(OutDir)" />
+    <Copy SourceFiles="$([System.IO.Path]::Combine($(NugetPackageRoot), $(FSharpCoreXmlDocRelativePath)))" DestinationFolder="$(OutDir)" />
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/FSharp.Common.fsproj:67]: Please don't change unrelated lines.

File: ReSharper.FSharp/src/FSharp.Common/src/Checker/FSharpCheckerService.fs [hunk 30]
```diff
@@ -27,9 +27,8 @@ module FSharpCheckerService =
 
 
 [<ShellComponent; AllowNullLiteral>]
-type FSharpCheckerService
-        (lifetime: Lifetime, logger: ILogger, onSolutionCloseNotifier: OnSolutionCloseNotifier,
-         settingsStore: ISettingsStore, settingsSchema: SettingsSchema, reactorMonitor: IFcsReactorMonitor) =
+type FSharpCheckerService(lifetime: Lifetime, logger: ILogger, onSolutionCloseNotifier: OnSolutionCloseNotifier,
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Checker/FSharpCheckerService.fs]: Why this file is changes in this PR?

File: ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsProjectProvider.fs [hunk 31]
```diff
@@ -52,11 +52,10 @@ module FcsProjectProvider =
         ProjectModelChangeType.REFERENCE_TARGET
 
 [<SolutionComponent>]
-type FcsProjectProvider
-        (lifetime: Lifetime, solution: ISolution, changeManager: ChangeManager, checkerService: FSharpCheckerService,
-         fcsProjectBuilder: FcsProjectBuilder, scriptFcsProjectProvider: IScriptFcsProjectProvider,
-         scheduler: ISolutionLoadTasksScheduler, fsFileService: IFSharpFileService, psiModules: IPsiModules,
-         locks: IShellLocks, logger: ILogger) as this =
+type FcsProjectProvider(lifetime: Lifetime, solution: ISolution, changeManager: ChangeManager,
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsProjectProvider.fs]: Same as above.

File: ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsProjectProvider.fs [hunk 32]
```diff
@@ -307,9 +306,8 @@ type FcsProjectProvider
 
 /// Invalidates psi caches when a non-F# project is built and FCS cached resolve results become stale
 [<SolutionComponent>]
-type OutputAssemblyChangeInvalidator
-        (lifetime: Lifetime, outputAssemblies: OutputAssemblies, daemon: IDaemon, psiFiles: IPsiFiles,
-         fcsProjectProvider: IFcsProjectProvider, scheduler: ISolutionLoadTasksScheduler) =
+type OutputAssemblyChangeInvalidator(lifetime: Lifetime, outputAssemblies: OutputAssemblies, daemon: IDaemon,
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsProjectProvider.fs]: Same as above.

And perhaps more to come...

**Review by auduchinok — COMMENTED** [review]
Please rebase the branch. The diff currently contains a lot of existing changes.

File: ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsProjectProvider.fs (hunk 32 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsProjectProvider.fs]: don't worry! The branch has already been rebased.

**Comment by auduchinok** [comment]
@DedSec256 Congrats on having this merged!
We should probably do some cleanup here and there, but it's OK to do it separately.

**Comment by varon** [comment]
THIS IS AMAZING. WELL DONE GUYS!
image

**Comment by jam40jeff** [comment]
This seems as if it may have broken the functionality of some type providers, such as FSharp.Configuration, that need to read a project's App.config or web.config file.  The config file cannot be found anymore.  This works with the older version of Rider before I upgraded.

**Comment by DedSec256** [comment]
@jam40jeff, Hello! 
Can you please provide more information about your problem on our [issue tracker](https://youtrack.jetbrains.com)? 
And it would help us a lot if you could attach a project that reproduces the problem, and logs (Help -> Collect logs)

**Comment by jam40jeff** [comment]
@DedSec256 Thanks for the workaround.  Reverting back to in-proc type provider hosting solves the issue.  I will open a ticket with a repro when I have time.  Thanks for your help.

**Comment by auduchinok** [comment]
@jam40jeff Could you please share some info that could help us reproduce it? What OS and what type provider package version you use? What MSBuild version is chosen in Settings -> Build and Toolset?
